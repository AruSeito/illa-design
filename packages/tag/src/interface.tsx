import { HTMLAttributes } from "react"

export type ColorScheme =
  string
  | "whiteAlpha"
  | "blackAlpha"
  | "gray"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "teal"
  | "blue"
  | "cyan"
  | "purple"
  | "pink"
  | "linkedin"
  | "facebook"
  | "messenger"
  | "whatsapp"
  | "twitter"
  | "telegram"

export interface TagProps extends HTMLAttributes<HTMLDivElement> {
  colorScheme?: ColorScheme
  size?: "small" | "middle" | "large"
  variant?: "outline" | "solid" | "subtle"
  visible?: boolean
  closable?: boolean
}