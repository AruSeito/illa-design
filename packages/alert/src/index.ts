export * from "./alert"
export * from "./interface"
