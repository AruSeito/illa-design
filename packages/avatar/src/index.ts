export * from "./avatar"
export * from "./interface"
export * from "./avatar-group"