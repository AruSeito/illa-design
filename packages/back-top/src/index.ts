export * from "./backtop"
export * from "./interface"
export * from "./style"
