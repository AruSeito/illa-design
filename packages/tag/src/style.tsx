import { css } from "@emotion/react"
import { TagColorScheme, TagVariant } from "./interface"
import { SerializedStyles } from "@emotion/serialize"
import { globalColor, illaPrefix } from "@illa-design/theme"

export const tagContainer = css`
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  flex-direction: row;
  align-items: center;
`

export function applyTagSizeLarge(variant: TagVariant): SerializedStyles {
  return css`
    padding: ${variant == "outline" ? "4px 7px" : "5px 8px"};
  `
}

export function applyTagSizeMedium(variant: TagVariant): SerializedStyles {
  return css`
    padding: ${variant == "outline" ? "2px 7px" : "3px 8px"};
  `
}

export function applyTagSizeSmall(variant: TagVariant): SerializedStyles {
  return css`
    padding: ${variant == "outline" ? "0px 7px" : "1px 8px"};
  `
}

export const leftIcon = css`
  width: 12px;
  height: 12px;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  margin-right: 4px;
`

export const closeIcon = css`
  margin-left: 4px;
`

export const colors: TagColorScheme[] = [
  "white", "blackAlpha", "gray", "grayBlue", "red", "orange", "yellow", "green", "blue", "cyan", "purple", "techBlue",
]

export function tagOutlinePrepare(color: TagColorScheme): SerializedStyles {
  if (color == "gray") {
    return css`
      border-radius: 1px;
      border: solid 1px ${globalColor(`--${illaPrefix}-${color}-08`)};
      color: ${globalColor(`--${illaPrefix}-${color}-02`)};
    `
  } else {
    return css`
      border-radius: 1px;
      border: solid 1px ${globalColor(`--${illaPrefix}-${color}-01`)};
      color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    `
  }
}

export function tagFillPrepare(color: TagColorScheme): SerializedStyles {
  return css`
    background-color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    color: ${globalColor(`--${illaPrefix}-white-01`)};
    border-radius: 1px;
  `
}

export function tagLightPrepare(color: TagColorScheme): SerializedStyles {
  if (color == "gray") {
    return css`
      border-radius: 1px;
      background-color: ${globalColor(`--${illaPrefix}-${color}-08`)};
      color: ${globalColor(`--${illaPrefix}-${color}-02`)};
    `
  } else {
    return css`
      border-radius: 1px;
      background-color: ${globalColor(`--${illaPrefix}-${color}-06`)};
      color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    `
  }
}

export function tagFillNormal(color: Extract<TagColorScheme, string>): SerializedStyles {
  return css`
    border-radius: 1px;
    color: ${globalColor(`--${illaPrefix}-white-01`)};
    background-color: ${color};
  `
}

export function tagOutlineNormal(color: Extract<TagColorScheme, string>): SerializedStyles {
  return css`
    border-radius: 1px;
    color: ${color};
    border: solid 1px ${color};
  `
}