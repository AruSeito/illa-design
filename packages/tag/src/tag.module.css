.red {
    font-size: 1000px;
}