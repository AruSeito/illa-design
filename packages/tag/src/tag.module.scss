.tag-container {
    display: inline-flex;
    flex-direction: row;
    align-items: center;
}

.tag-outline-$color {
    border: solid 1px $color;
}