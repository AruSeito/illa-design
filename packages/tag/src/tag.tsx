/** @jsxImportSource @emotion/react */
import * as React from "react"
import { forwardRef } from "react"
import { TagColorScheme, TagProps } from "./interface"
import { css } from "@emotion/react"
import { CloseIcon } from "@illa-design/icon"
import { SerializedStyles } from "@emotion/serialize"
import { globalColor, illaPrefix } from "@illa-design/theme"
import { omit } from "@illa-design/system"

// style
const tagContainer = css`
  display: inline-flex;
  flex-direction: row;
  align-items: center;
`

const tagSizeLarge = css`
  padding: 5px 8px;
`

const tagSizeMedium = css`
  padding: 3px 8px;
`

const tagSizeSmall = css`
  padding: 1px 8px;
`

const leftIcon = css`
  width: 12px;
  height: 12px;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  margin-right: 4px;
`

const closeIcon = css`
  margin-left: 4px;
`

const colors: TagColorScheme[] = [
  "white", "blackAlpha", "gray", "grayBlue", "red", "orange", "yellow", "green", "blue", "cyan", "purple",
]

function tagOutlinePrepare(color: TagColorScheme): SerializedStyles {
  if (color == "gray") {
    return css`
      border-radius: 1px;
      border: solid 1px ${globalColor(`--${illaPrefix}-${color}-08`)};
      color: ${globalColor(`--${illaPrefix}-${color}-02`)};
    `
  } else {
    return css`
      border-radius: 1px;
      border: solid 1px ${globalColor(`--${illaPrefix}-${color}-01`)};
      color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    `
  }
}

function tagFillPrepare(color: TagColorScheme): SerializedStyles {
  return css`
    background-color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    color: ${globalColor(`--${illaPrefix}-white-01`)};
    border-radius: 1px;
  `
}

function tagLightPrepare(color: TagColorScheme): SerializedStyles {
  if (color == "gray") {
    return css`
      border-radius: 1px;
      background-color: ${globalColor(`--${illaPrefix}-${color}-08`)};
      color: ${globalColor(`--${illaPrefix}-${color}-02`)};
    `
  } else {
    return css`
      border-radius: 1px;
      background-color: ${globalColor(`--${illaPrefix}-${color}-06`)};
      color: ${globalColor(`--${illaPrefix}-${color}-01`)};
    `
  }
}

function tagFillNormal(color: Extract<TagColorScheme, string>): SerializedStyles {
  return css`
    border-radius: 1px;
    color: ${globalColor(`--${illaPrefix}-white-01`)};
    background-color: ${color};
  `
}

function tagOutlineNormal(color: Extract<TagColorScheme, string>): SerializedStyles {
  return css`
    border-radius: 1px;
    color: ${color};
    border: solid 1px ${color};
  `
}

export const Tag = forwardRef<HTMLDivElement, TagProps>((props, ref) => {

  const {
    visible = true,
    colorScheme = "gray",
    size = "small",
    variant = "light",
    closable = false,
    ...rest
  } = props

  const otherProps = omit(rest, ["onClose", "icon"])

  let variantCss: SerializedStyles

  // variant
  if (colors.includes(colorScheme)) {
    switch (variant) {
      case "light": {
        variantCss = tagLightPrepare(colorScheme)
        break
      }
      case "fill": {
        variantCss = tagFillPrepare(colorScheme)
        break
      }
      case "outline": {
        variantCss = tagOutlinePrepare(colorScheme)
        break
      }
    }
  } else {
    switch (variant) {
      case "light":
      case "fill": {
        variantCss = tagFillNormal(colorScheme)
        break
      }
      case "outline": {
        variantCss = tagOutlineNormal(colorScheme)
        break
      }
    }
  }

  // size
  let sizeCss: SerializedStyles
  switch (size) {
    case "small": {
      sizeCss = tagSizeSmall
      break
    }
    case "medium": {
      sizeCss = tagSizeMedium
      break
    }
    case "large": {
      sizeCss = tagSizeLarge
      break
    }
  }

  const finalStyle = css`
    ${tagContainer};
    ${variantCss};
    ${sizeCss};
  `

  return visible ? <div ref={ref} {...otherProps}>
    <div css={finalStyle}>
      {props.icon && <span css={leftIcon}>{props.icon}</span>}
      <span css={css`
        font-size: 14px;
        line-height: 22px;
      `}>{props.children}</span>
      {props.closable && <CloseIcon measure="7px" css={closeIcon} onClick={() => {
        if (props.onClose != undefined) {
          props.onClose()
        }
      }} />}
    </div>
  </div> : null
})