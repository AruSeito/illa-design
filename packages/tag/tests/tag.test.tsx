import { Tag } from "../src"
import { create } from "react-test-renderer"
import { BsFacebook } from "react-icons/bs"

test("Tag renders with text", () => {
  const node = create(<Tag className="test-with-text">Tag renders with text</Tag>)
  expect(node.root.findByType(Tag).props.className).toBe("test-with-text")
})

test("Tag renders with icon", () => {
  const node = create(<Tag className="test-with-icon" icon={<BsFacebook />} />)
  expect(node.root.findByType(Tag).props.className).toBe("test-with-icon")
})

test("Tag renders invisible", () => {
  const node = create(
    <Tag className="test-invisible" visible={false} />,
  )
  expect(node.root.findByType(Tag).instance).toBe(null)
})

test("Tag renders with close icon", () => {
  const node = create(
    <Tag className="test-with-close-icon" closable />,
  )
  expect(node.root.findByType(Tag).props.className).toBe("test-with-close-icon")
})

test("Tag renders close event", () => {
  const onClose = jest.fn()
  const node = create(
    <Tag className="test-invisible" visible={false} onClose={onClose} />,
  )
  node.root.findByProps({ className: "test-invisible" }).props.onClose()
  expect(onClose).toBeCalled()
})