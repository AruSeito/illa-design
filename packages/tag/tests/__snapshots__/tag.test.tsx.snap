// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders with color and variant 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  background-color: #ebebeb;
  color: #1f1f1f;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 2`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  background-color: #a3f0d4;
  color: #007a41;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 3`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  color: #fff;
  background-color: #123321;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 4`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #000;
  color: #fff;
  border-radius: 1px;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 5`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #007a41;
  color: #fff;
  border-radius: 1px;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 6`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  color: #fff;
  background-color: #123321;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 7`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  border: solid 1px #ebebeb;
  color: #1f1f1f;
  padding: 0px 7px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 8`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  border: solid 1px #007a41;
  color: #007a41;
  padding: 0px 7px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with color and variant 9`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  color: #123321;
  border: solid 1px #123321;
  padding: 0px 7px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with different size 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  background-color: #ebebeb;
  color: #1f1f1f;
  padding: 1px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with different size 2`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  background-color: #ebebeb;
  color: #1f1f1f;
  padding: 3px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;

exports[`Tag renders with different size 3`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1px;
  background-color: #ebebeb;
  color: #1f1f1f;
  padding: 5px 8px;
}

.emotion-1 {
  font-size: 14px;
  line-height: 22px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      Hello World
    </span>
  </div>
</DocumentFragment>
`;
